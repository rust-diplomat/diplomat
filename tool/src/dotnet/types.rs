use core::fmt;
use diplomat_core::ast;
use diplomat_core::Env;
use heck::ToUpperCamelCase;

pub fn gen_type_name_to_string(
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
) -> Result<String, fmt::Error> {
    let mut s = String::new();
    gen_type_name(typ, in_path, env, &mut s)?;
    Ok(s)
}

pub fn gen_type_name(
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
    out: &mut dyn fmt::Write,
) -> fmt::Result {
    match typ {
        ast::TypeName::Named(path_type) | ast::TypeName::SelfType(path_type) => {
            write!(out, "{}", path_type.resolve(in_path, env).name())
        }

        ast::TypeName::Box(underlying) => gen_type_name(underlying.as_ref(), in_path, env, out),

        ast::TypeName::Reference(.., underlying) => {
            gen_type_name(underlying.as_ref(), in_path, env, out)
        }

        ast::TypeName::Option(underlying) => gen_type_name(underlying.as_ref(), in_path, env, out),

        ast::TypeName::Result(..) => {
            write!(
                out,
                "{}{}",
                in_path.elements.join("_").to_upper_camel_case(),
                name_for_type(typ)
            )
        }

        ast::TypeName::Primitive(prim) => {
            write!(out, "{}", type_name_for_prim(prim))
        }

        ast::TypeName::Writeable => {
            write!(out, "DiplomatWriteable")
        }

        ast::TypeName::StrReference(
            _,
            ast::StringEncoding::UnvalidatedUtf8 | ast::StringEncoding::Utf8,
        ) => {
            write!(out, "string")
        }

        ast::TypeName::StrReference(_, ast::StringEncoding::UnvalidatedUtf16) => {
            write!(out, "ushort[]")
        }

        ast::TypeName::PrimitiveSlice(.., prim) => {
            write!(out, "{}[]", type_name_for_prim(prim))
        }

        ast::TypeName::Unit => {
            write!(out, "void")
        }
        &_ => unreachable!("unknown AST/HIR variant"),
    }
}

pub fn type_name_for_prim(prim: &ast::PrimitiveType) -> &str {
    match prim {
        ast::PrimitiveType::i8 => "sbyte",
        ast::PrimitiveType::u8 => "byte",
        ast::PrimitiveType::i16 => "short",
        ast::PrimitiveType::u16 => "ushort",
        ast::PrimitiveType::i32 => "int",
        ast::PrimitiveType::u32 => "uint",
        ast::PrimitiveType::i64 => "long",
        ast::PrimitiveType::u64 => "ulong",
        ast::PrimitiveType::i128 => "Int128",
        ast::PrimitiveType::u128 => "UInt128",
        ast::PrimitiveType::isize => "nint",
        ast::PrimitiveType::usize => "nuint",
        ast::PrimitiveType::f32 => "float",
        ast::PrimitiveType::f64 => "double",
        ast::PrimitiveType::bool => "bool",
        ast::PrimitiveType::char => "uint",
    }
}

/// Generates a struct name that uniquely identifies the given type.
pub fn name_for_type(typ: &ast::TypeName) -> ast::Ident {
    match typ {
        ast::TypeName::Named(name) | ast::TypeName::SelfType(name) => {
            name.path.elements.last().unwrap().clone()
        }
        ast::TypeName::Box(underlying) => {
            ast::Ident::from(format!("Box{}", name_for_type(underlying)))
        }
        ast::TypeName::Reference(_, ast::Mutability::Mutable, underlying) => {
            ast::Ident::from(format!("RefMut{}", name_for_type(underlying)))
        }
        ast::TypeName::Reference(_, ast::Mutability::Immutable, underlying) => {
            ast::Ident::from(format!("Ref{}", name_for_type(underlying)))
        }
        ast::TypeName::Primitive(prim) => ast::Ident::from(prim.to_string().to_upper_camel_case()),
        ast::TypeName::Option(underlying) => {
            ast::Ident::from(format!("Opt{}", name_for_type(underlying)))
        }
        ast::TypeName::Result(ok, err, _) => {
            ast::Ident::from(format!("Result{}{}", name_for_type(ok), name_for_type(err)))
        }
        ast::TypeName::Writeable => ast::Ident::from("Writeable"),
        ast::TypeName::StrReference(
            _,
            ast::StringEncoding::UnvalidatedUtf8 | ast::StringEncoding::Utf8,
        ) => ast::Ident::from("StrRef8"),
        ast::TypeName::StrReference(_, ast::StringEncoding::UnvalidatedUtf16) => {
            ast::Ident::from("RefMutPrimSliceU16")
        }
        ast::TypeName::PrimitiveSlice(_, ast::Mutability::Mutable, prim) => ast::Ident::from(
            format!("RefMutPrimSlice{}", prim.to_string().to_upper_camel_case()),
        ),
        ast::TypeName::PrimitiveSlice(_, ast::Mutability::Immutable, prim) => ast::Ident::from(
            format!("RefPrimSlice{}", prim.to_string().to_upper_camel_case()),
        ),
        ast::TypeName::Unit => ast::Ident::from("Void"),
        &_ => unreachable!("unknown AST/HIR variant"),
    }
}
