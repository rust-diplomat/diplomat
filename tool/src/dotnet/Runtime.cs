// Automatically generated by Diplomat

using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;

namespace {{NAMESPACE}}.Diplomat;

#nullable enable

[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
delegate void WriteFlush(IntPtr self);

[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
[return: MarshalAs(UnmanagedType.U1)]
delegate bool WriteGrow(IntPtr self, nuint capacity);

[Serializable]
[StructLayout(LayoutKind.Sequential)]
public struct DiplomatWrite : IDisposable
{
    // About the current approach:
    // Ideally DiplomatWrite should wrap the native string type and grows/writes directly into the internal buffer.
    // However, there is no native string type backed by UTF-8 in dotnet frameworks.
    // Alternative could be to provide Diplomat's own `Utf8String` type, but this is not trivial and mostly useless
    // on its own because all other dotnet/C# APIs are expecting the standard UTF-16 encoded string/String type.
    // API is expectedâ€¯to become overall less ergonomic by forcing users to explicitly convert between UTF-16 and UTF-8 anyway in such case.
    // It could be useful for copy-free interactions between Diplomat-generated API though.
    // Also, there is no way to re-use the unmanaged buffer as-is to get a managed byte[]
    // (hence `ToUtf8Bytes` is copying the internal buffer from unmanaged to managed memory).
    // It's encouraged to enable memoization when applicable (#129).

    IntPtr context;
    IntPtr buf;
    nuint len;
    nuint cap;
    readonly IntPtr flush;
    readonly IntPtr grow;

    public DiplomatWrite()
    {
        WriteFlush flushFunc = Flush;
        WriteGrow growFunc = Grow;

        IntPtr flushFuncPtr = Marshal.GetFunctionPointerForDelegate(flushFunc);
        IntPtr growFuncPtr = Marshal.GetFunctionPointerForDelegate(growFunc);
        
        // flushFunc and growFunc are managed objects and might be disposed of by the garbage collector.
        // To prevent this, we make the context hold the references and protect the context itself
        // for automatic disposal by moving it behind a GCHandle.
        DiplomatWriteContext ctx = new DiplomatWriteContext();        
        ctx.flushFunc = flushFunc;
        ctx.growFunc = growFunc;
        GCHandle ctxHandle = GCHandle.Alloc(ctx);

        context = GCHandle.ToIntPtr(ctxHandle);
        buf = Marshal.AllocHGlobal(64);
        len = 0;
        cap = 64;
        flush = flushFuncPtr;
        grow = growFuncPtr;
    }

    public byte[] ToUtf8Bytes()
    {
        if (len > int.MaxValue)
        {
            throw new IndexOutOfRangeException("DiplomatWrite buffer is too big");
        }
        byte[] managedArray = new byte[(int)len];
        Marshal.Copy(buf, managedArray, 0, (int)len);
        return managedArray;
    }

    public string ToUnicode()
    {
#if NET6_0_OR_GREATER
        if (len > int.MaxValue)
        {
            throw new IndexOutOfRangeException("DiplomatWrite buffer is too big");
        }
        return Marshal.PtrToStringUTF8(buf, (int) len);
#else
        byte[] utf8 = ToUtf8Bytes();
        return DiplomatUtils.Utf8ToString(utf8);
#endif
    }

    public void Dispose()
    {
        if (buf != IntPtr.Zero)
        {
            Marshal.FreeHGlobal(buf);
            buf = IntPtr.Zero;
        }

        if (context != IntPtr.Zero)
        {
            GCHandle.FromIntPtr(context).Free();
            context = IntPtr.Zero;
        }
    }

    private static void Flush(IntPtr self)
    {
        // Nothing to do
    }

    [return: MarshalAs(UnmanagedType.U1)]
    private unsafe static bool Grow(IntPtr write, nuint capacity)
    {
        if (write == IntPtr.Zero)
        {
            return false;
        }
        DiplomatWrite* self = (DiplomatWrite*)write;

        nuint newCap = capacity;
        if (newCap > int.MaxValue)
        {
            return false;
        }

        IntPtr newBuf;
        try
        {
            newBuf = Marshal.AllocHGlobal((int)newCap);
        }
        catch (OutOfMemoryException)
        {
            return false;
        }

        Buffer.MemoryCopy((void*)self->buf, (void*)newBuf, newCap, self->cap);
        Marshal.FreeHGlobal(self->buf);
        self->buf = newBuf;
        self->cap = newCap;

        return true;
    }
}

internal struct DiplomatWriteContext
{
    internal WriteFlush flushFunc;
    internal WriteGrow growFunc;
}

internal static class DiplomatUtils
{
    internal static byte[] StringToUtf8(string s)
    {
        int size = Encoding.UTF8.GetByteCount(s);
        byte[] buf = new byte[size];
        Encoding.UTF8.GetBytes(s, 0, s.Length, buf, 0);
        return buf;
    }

    internal static string Utf8ToString(byte[] utf8)
    {
        char[] chars = new char[utf8.Length];
        Encoding.UTF8.GetChars(utf8, 0, utf8.Length, chars, 0);
        return new string(chars);
    }
}

public class DiplomatOpaqueException : Exception
{
    public DiplomatOpaqueException() : base("The FFI function failed with an opaque error") { }
}
