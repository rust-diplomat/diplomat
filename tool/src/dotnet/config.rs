use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct LibraryConfig {
    pub namespace: String,
    pub native_lib: String,
    #[serde(default)]
    pub usings: Vec<String>,
    #[serde(default)]
    pub exceptions: ExceptionsConfig,
    #[serde(default)]
    pub properties: PropertiesConfig,
}

impl LibraryConfig {
    pub fn default() -> LibraryConfig {
        LibraryConfig {
            namespace: "Interop".to_owned(),
            native_lib: "rust".to_owned(),
            usings: Vec::new(),
            exceptions: ExceptionsConfig::default(),
            properties: PropertiesConfig::default(),
        }
    }
}

#[derive(Debug, Deserialize, Default)]
pub struct PropertiesConfig {
    #[serde(default)]
    pub setters_prefix: Option<String>,
    #[serde(default)]
    pub getters_prefix: Option<String>,
}

#[derive(Debug, Deserialize, Default)]
pub struct ExceptionsConfig {
    #[serde(default)]
    pub trim_suffix: String,
    #[serde(default)]
    pub error_message_method: Option<String>,
}
