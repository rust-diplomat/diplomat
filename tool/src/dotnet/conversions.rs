use std::fmt::{self, Write as _};

use diplomat_core::{ast, Env};

use super::types::{gen_type_name_to_string, type_name_for_prim};
use crate::util::CodeWriter;

pub struct SliceParam {
    pub array_var_name: String,
    pub ptr_var_name: String,
    pub length_var_name: String,
    pub underlying_type: ast::PrimitiveType,
}

impl SliceParam {
    pub fn new(name: String, underlying_type: ast::PrimitiveType) -> Self {
        let ptr_var_name = format!("{}Ptr", name);
        let length_var_name = format!("{}Length", name);
        Self {
            array_var_name: name,
            ptr_var_name,
            length_var_name,
            underlying_type,
        }
    }

    pub fn open_fixed_block(&self, out: &mut CodeWriter) -> fmt::Result {
        writeln!(
            out,
            "fixed ({}* {} = {})",
            type_name_for_prim(&self.underlying_type),
            self.ptr_var_name,
            self.array_var_name,
        )?;
        writeln!(out, "{{")?;
        out.indent();
        Ok(())
    }

    pub fn close_fixed_block(self, out: &mut CodeWriter) -> fmt::Result {
        out.dedent();
        writeln!(out, "}}")
    }
}

/// Writes the raw → idiomatic conversion code
///
/// Primitive type: nothing to do
/// Struct/opaque custom type: wrap pointer into a managed class
/// Enum: cast from raw enum
pub fn to_idiomatic_object<W: fmt::Write>(
    env: &Env,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    input_var_name: &str,
    out: &mut W,
) -> fmt::Result {
    match typ {
        ast::TypeName::Primitive(_) => out.write_str(input_var_name),
        ast::TypeName::Box(boxed) => {
            to_idiomatic_object(env, boxed.as_ref(), in_path, input_var_name, out)
        }
        ast::TypeName::Reference(.., reference) => {
            to_idiomatic_object(env, reference.as_ref(), in_path, input_var_name, out)
        }
        ast::TypeName::Option(opt) => {
            to_idiomatic_object(env, opt.as_ref(), in_path, input_var_name, out)
        }
        _ => {
            let name = gen_type_name_to_string(typ, in_path, env)?;
            match typ {
                ast::TypeName::Named(path_type) => match path_type.resolve(in_path, env) {
                    ast::CustomType::Struct(_) | ast::CustomType::Opaque(_) => {
                        write!(out, "new {name}({input_var_name})")
                    }
                    ast::CustomType::Enum(_) => {
                        write!(out, "({name}){input_var_name}")
                    }
                },
                other => panic!("expected named type name, found `{}`", other),
            }
        }
    }
}

/// Writes the idiomatic → raw conversion code
///
/// Primitive type: nothing to do
/// Struct/opaque custom type: extract raw representation using `.AsFFI()`
/// Enum: cast from idiomatic enum
pub fn to_raw_object<W: fmt::Write>(
    env: &Env,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    input_var_name: &str,
    out: &mut W,
) -> fmt::Result {
    match typ {
        ast::TypeName::Primitive(_) => out.write_str(input_var_name),
        ast::TypeName::Box(boxed) => {
            to_raw_object(env, boxed.as_ref(), in_path, input_var_name, out)
        }
        ast::TypeName::Reference(.., reference) => {
            to_raw_object(env, reference.as_ref(), in_path, input_var_name, out)
        }
        ast::TypeName::Option(opt) => {
            to_raw_object(env, opt.as_ref(), in_path, input_var_name, out)
        }
        _ => {
            let name = gen_type_name_to_string(typ, in_path, env)?;
            match typ {
                ast::TypeName::Named(path_type) => match path_type.resolve(in_path, env) {
                    ast::CustomType::Struct(_) | ast::CustomType::Opaque(_) => {
                        write!(out, "{input_var_name}.AsFFI()")
                    }
                    ast::CustomType::Enum(_) => {
                        write!(out, "(Raw.{name}){input_var_name}")
                    }
                },
                other => panic!("expected named type name, found `{}`", other),
            }
        }
    }
}
