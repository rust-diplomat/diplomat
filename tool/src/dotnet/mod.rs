use std::collections::HashMap;
use std::fmt;
use std::fmt::Write;
use std::fs;
use std::path::PathBuf;

use colored::*;
use diplomat_core::Env;

use self::config::LibraryConfig;
use crate::util::CodeWriter;
use crate::util::SetOfAstTypes;

mod config;
mod conversions;
mod idiomatic;
mod raw;
mod types;
mod util;

static RUNTIME_CS: &str = include_str!("Runtime.cs");

const INDENTATION: &str = "    ";
const SCOPE_OPENING: &str = "{";
const SCOPE_CLOSING: &str = "}";

pub fn gen_bindings(
    env: &Env,
    library_config_path: &Option<PathBuf>,
    docs_url_gen: &diplomat_core::ast::DocsUrlGenerator,
    outs: &mut HashMap<String, String>,
) -> fmt::Result {
    let mut library_config = LibraryConfig::default();
    if let Some(path) = library_config_path {
        // Should be fine, we've already verified the path
        let contents = fs::read_to_string(path).unwrap();
        match toml::from_str(&contents) {
            Ok(config) => library_config = config,
            Err(err) => {
                eprintln!(
                    "{} Unable to parse library configuration file: {path:?}\n{err}",
                    "Error:".red().bold(),
                );
                std::process::exit(1);
            }
        }
    }

    let diplomat_runtime_out = outs
        .entry("DiplomatRuntime.cs".to_owned())
        .or_insert_with(String::new);
    write!(
        diplomat_runtime_out,
        "{}",
        RUNTIME_CS.replace("{{NAMESPACE}}", &library_config.namespace)
    )?;

    let all_types = crate::util::get_all_custom_types(env);
    let mut errors = SetOfAstTypes::default();

    {
        // Raw API generation pass

        let mut results = SetOfAstTypes::default();

        for (in_path, typ) in &all_types {
            let mut out_buf = String::new();
            let mut out = CodeWriter::new(&mut out_buf, INDENTATION, SCOPE_OPENING, SCOPE_CLOSING);
            raw::gen_header(&library_config, &mut out)?;
            raw::gen(
                env,
                &library_config,
                &mut results,
                &mut errors,
                typ,
                in_path,
                docs_url_gen,
                &mut out,
            )?;
            outs.insert(format!("Raw{}.cs", typ.name()), out_buf)
                .and_then::<String, _>(|_| panic!("file created twice: Raw{}.cs", typ.name()));
        }

        for (in_path, typ) in results {
            let result_name = types::gen_type_name_to_string(typ, &in_path, env)?;
            let mut out_buf = String::new();
            let mut out = CodeWriter::new(&mut out_buf, INDENTATION, SCOPE_OPENING, SCOPE_CLOSING);
            raw::gen_header(&library_config, &mut out)?;
            raw::gen_result(typ, &in_path, env, &mut out)?;
            outs.insert(format!("Raw{}.cs", result_name), out_buf)
                .and_then::<String, _>(|_| panic!("file created twice: Raw{}.cs", result_name));
        }
    }

    {
        // Idiomatic API generation pass

        for (in_path, typ) in &all_types {
            let mut out_buf = String::new();
            let mut out = CodeWriter::new(&mut out_buf, INDENTATION, SCOPE_OPENING, SCOPE_CLOSING);
            idiomatic::gen_header(&library_config, &mut out)?;
            idiomatic::gen(typ, in_path, env, &library_config, docs_url_gen, &mut out)?;
            outs.insert(format!("{}.cs", typ.name()), out_buf)
                .and_then::<String, _>(|_| panic!("file created twice: {}.cs", typ.name()));
        }

        for (in_path, typ) in errors {
            let idiomatic::ExceptionCtx { name, .. } =
                idiomatic::error_type_to_exception_name(env, &library_config, typ, &in_path)?;
            let mut out_buf = String::new();
            let mut out = CodeWriter::new(&mut out_buf, INDENTATION, SCOPE_OPENING, SCOPE_CLOSING);
            idiomatic::gen_header(&library_config, &mut out)?;
            idiomatic::gen_exception(env, &library_config, typ, &in_path, &mut out)?;
            outs.insert(format!("{}.cs", name), out_buf)
                .and_then::<String, _>(|_| panic!("file created twice: {}.cs", name));
        }
    }

    Ok(())
}
