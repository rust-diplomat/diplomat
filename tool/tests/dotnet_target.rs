use diplomat_tool::dotnet::*;

macro_rules! test_file {
    ($($file:tt)*) => {
        let parsed: syn::File = syn::parse_quote! { $($file)* };
        let custom_types = diplomat_core::ast::File::from(&parsed);
        let env = custom_types.all_types();

        let mut out_texts = std::collections::HashMap::new();

        gen_bindings(&env, None, &Default::default(), &mut out_texts).unwrap();
        out_texts.remove("DiplomatRuntime.cs");

        for out in out_texts.keys() {
            insta::with_settings!({ snapshot_suffix => out.clone() }, {
                insta::assert_snapshot!(out_texts.get(out).unwrap())
            });
        }
    }
}

macro_rules! test_file_using_library_config {
    ($($file:tt)*) => {
        let parsed: syn::File = syn::parse_quote! { $($file)* };
        let custom_types = diplomat_core::ast::File::from(&parsed);
        let env = custom_types.all_types();

        let mut out_texts = std::collections::HashMap::new();

        use std::path::PathBuf;
        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        path.push("tests/configs/dotnet_example.toml");
        gen_bindings(&env, Some(&path), &Default::default(), &mut out_texts).unwrap();
        out_texts.remove("DiplomatRuntime.cs");

        for out in out_texts.keys() {
            insta::with_settings!({ snapshot_suffix => out.clone() }, {
                insta::assert_snapshot!(out_texts.get(out).unwrap())
            });
        }
    }
}

#[test]
fn generation_using_default_config() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyLibError;

            impl MyLibError {
                pub fn to_display(out: &mut DiplomatWrite) {
                    unimplemented!()
                }
            }

            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn new_slice(v: &[f64]) -> Result<Box<MyStruct>, Box<MyLibError>> {
                    unimplemented!()
                }

                pub fn set_slice(&mut self, new_slice: &[f64]) -> Result<(), Box<MyLibError>> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn generation_using_library_config() {
    test_file_using_library_config! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyLibError;

            impl MyLibError {
                pub fn to_display(&self, out: &mut DiplomatWrite) {
                    unimplemented!()
                }
            }

            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn new_slice(v: &[f64]) -> Result<Box<MyStruct>, Box<MyLibError>> {
                    unimplemented!()
                }

                pub fn set_slice(&mut self, new_slice: &[f64]) -> Result<(), Box<MyLibError>> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn setters_getters_properties() {
    test_file_using_library_config! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct Foo;

            #[diplomat::opaque]
            struct Bar;

            impl Bar {
                pub fn get_foo(&self) -> Result<Box<Foo>, ()> {
                    unimplemented!()
                }

                pub fn set_foo(&mut self, foo: &Foo) -> Result<(), ()> {
                    unimplemented!()
                }

                pub fn get_name(&self, out: &mut DiplomatWrite) -> Result<(), ()> {
                    unimplemented!()
                }

                pub fn set_name(&mut self, new_name: &DiplomatStr) -> Result<(), ()> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn cross_module_struct_fields() {
    test_file! {
        #[diplomat::bridge]
        mod mod1 {
            use super::mod2::Bar;

            struct Foo {
                x: Bar,
            }
        }

        #[diplomat::bridge]
        mod mod2 {
            use super::mod1::Foo;

            struct Bar {
                y: Box<Foo>,
            }
        }
    }
}

#[test]
fn cross_module_struct_methods() {
    test_file! {
        #[diplomat::bridge]
        mod mod1 {
            use super::mod2::Bar;

            #[diplomat::opaque]
            struct Foo;

            impl Foo {
                pub fn to_bar(&self) -> Bar {
                    unimplemented!()
                }
            }
        }

        #[diplomat::bridge]
        mod mod2 {
            use super::mod1::Foo;

            struct Bar {
                y: Box<Foo>,
            }
        }
    }
}

#[test]
fn enum_documentation() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            /// Documentation for MyEnum.
            ///
            /// Some remarks about this struct.
            #[diplomat::rust_link(foo::MyEnum, Enum)]
            enum MyEnum {
                /// All about A.
                A,
                /// All about B.
                #[diplomat::rust_link(foo::MyEnum::B, EnumVariant)]
                B,
                /// All about C.
                ///
                /// Some remarks about this variant.
                ///
                /// Even more remarks.
                C
            }
        }
    }
}

#[test]
fn simple_non_opaque_struct() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            struct MyStruct {
                a: u8,
                b: u8,
            }

            impl MyStruct {
                pub fn new(a: u8, b: u8) -> MyStruct {
                    unimplemented!()
                }

                pub fn get_a(&self) -> u8 {
                    unimplemented!()
                }

                pub fn set_b(&mut self, b: u8) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn simple_opaque_struct() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct(UnknownType);

            impl MyStruct {
                pub fn new(a: u8, b: u8) -> Box<MyStruct> {
                    unimplemented!()
                }

                pub fn get_a(&self) -> u8 {
                    unimplemented!()
                }

                pub fn set_b(&mut self, b: u8) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn method_taking_str() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct(UnknownType);

            impl MyStruct {
                pub fn new_str(v: &DiplomatStr) -> Box<MyStruct> {
                    unimplemented!()
                }

                pub fn set_str(&mut self, new_str: &DiplomatStr) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn method_taking_slice() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct(UnknownType);

            impl MyStruct {
                pub fn new_slice(v: &[f64]) -> Box<MyStruct> {
                    unimplemented!()
                }

                pub fn fill_slice(&self, v: &mut [f64]) {
                    unimplemented!()
                }

                pub fn set_slice(&mut self, new_slice: &[f64]) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn method_write_out() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct(UnknownType);

            impl MyStruct {
                pub fn write(&self, out: &mut DiplomatWrite) {
                    unimplemented!()
                }

                pub fn write_unit(&self, out: &mut DiplomatWrite) -> () {
                    unimplemented!()
                }

                pub fn write_result(&self, out: &mut DiplomatWrite) -> Result<(), u8> {
                    unimplemented!()
                }

                pub fn write_no_rearrange(&self, out: &mut DiplomatWrite) -> u8 {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn struct_documentation() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            /// Documentation for Foo.
            /// Second line.
            struct Foo {
                /// Documentation for x.
                x: u8,
            }

            impl Foo {
                /// Documentation for get_x.
                pub fn get_x(&self) -> u8 {
                    x
                }
            }
        }
    }
}

#[test]
fn enum_conversion() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            enum MyEnum {
                A, B, C
            }

            struct MyStruct {
                a: u8,
                b: MyEnum,
            }

            #[diplomat::opaque]
            struct Foo;

            impl Foo {
                pub fn get_struct(&self, a: u8, b: MyEnum) -> MyStruct {
                    MyStruct { a, b }
                }
            }
        }
    }
}

#[test]
fn option_conversion() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyOtherStruct;

            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn create(my_other_struct: Option<&MyOtherStruct>) -> Option<Box<MyStruct>> {
                    unimplemented!();
                }
            }
        }
    }
}

#[test]
fn pointer_types() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyOpaqueStruct(UnknownType);

            struct MyStruct<'a> {
                a: &'a MyOpaqueStruct,
            }

            impl<'a> MyStruct<'a> {
                pub fn new(foo: &'a MyOpaqueStruct, bar: &'a mut MyOpaqueStruct) -> MyStruct<'a> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn boolean_type() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            struct MyStruct {
                a: bool,
            }

            impl MyStruct {
                pub fn func(a: bool) -> bool {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn option_types() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyOpaqueStruct(UnknownType);

            struct MyStruct {
                a: Option<Box<MyOpaqueStruct>>,
            }
        }
    }
}

#[test]
fn result_types() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyOpaqueStruct(UnknownType);

            struct MyStruct {
                a: DiplomatResult<Box<MyOpaqueStruct>, u8>,
            }

            impl MyStruct {
                pub fn new() -> Result<MyStruct, u8> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn string_reference() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn new(v: &DiplomatStr) -> MyStruct {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn write_out() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn write(&self, to: &mut DiplomatWrite) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn unit_type() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn something(&self) -> () {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn error_handling() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct;

            #[diplomat::opaque]
            struct MyModuleOpaqueError;

            enum MyModuleError {
                BadInput,
                BadLuck,
            }

            impl MyStruct {
                pub fn foo(&self) -> Result<u32, MyModuleError> {
                    unimplemented!()
                }

                pub fn bar(&self) -> Result<DiplomatChar, MyModuleOpaqueError> {
                    unimplemented!()
                }

                pub fn baz(&self) -> Result<(), ()> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn almost_properties() {
    test_file_using_library_config! {
        #[diplomat::bridge]
        mod ffi {
            /// This should not contain any property
            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                /// This should not generate a property
                pub fn get_foo_by_key(&self, key: &DiplomatStr) -> Result<u64, ()> {
                    unimplemented!()
                }

                /// This should not generate a property
                pub fn set_foo_by_key(&self, key: &DiplomatStr, foo: i64) -> Result<(), ()> {
                    unimplemented!()
                }

                /// This should not generate a property
                pub fn get_str_by_key(&self, key: &DiplomatStr, writer: &mut DiplomatWrite) {
                    unimplemented!()
                }

                /// This should not generate a property
                pub fn set_str_by_key(&self, key: &DiplomatStr, s: &DiplomatStr) {
                    unimplemented!()
                }
            }
        }
    }
}
