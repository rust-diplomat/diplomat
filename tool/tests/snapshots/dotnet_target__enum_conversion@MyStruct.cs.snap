---
source: tool/tests/dotnet_target.rs
expression: out_texts.get(out).unwrap()

---
// Automatically generated by Diplomat

#pragma warning disable 0105
using System;
using System.Runtime.InteropServices;

using Interop.Diplomat;
#pragma warning restore 0105

namespace Interop;

#nullable enable

public partial class MyStruct
{
    private Raw.MyStruct _inner;

    public byte A
    {
        get
        {
            unsafe
            {
                return _inner.a;
            }
        }
        set
        {
            unsafe
            {
                _inner.a = value;
            }
        }
    }

    public MyEnum B
    {
        get
        {
            unsafe
            {
                return (MyEnum)_inner.b;
            }
        }
        set
        {
            unsafe
            {
                _inner.b = (Raw.MyEnum)value;
            }
        }
    }

    /// <summary>
    /// Creates a managed <c>MyStruct</c> from the raw representation.
    /// </summary>
    public unsafe MyStruct(Raw.MyStruct data)
    {
        _inner = data;
    }

    /// <summary>
    /// Returns a copy of the underlying raw representation.
    /// </summary>
    public Raw.MyStruct AsFFI()
    {
        return _inner;
    }
}

